File: vetly-front/types/vet.ts
/**
 * Vet types matching backend API responses
 */

export interface Vet {
  id: string;
  name: string;
  email: string;
  specialty: string;
  license_number: string;
  phone: string;
  address: string;
  city: string;
  coordinates_lat: number | null;
  coordinates_lng: number | null;
  hours: Record<string, DayHours> | null;
  description: string | null;
  image_url: string | null;
  is_on_call: boolean;
  is_verified: boolean;
  rating_average: number;
  reviews_count: number;
  created_at: string;
  updated_at: string;
}

export interface DayHours {
  open: string | null;
  close: string | null;
  closed: boolean;
}

export interface VetListResponse {
  items: Vet[];
  total: number;
  page: number;
  page_size: number;
}

export interface ServiceType {
  id: string;
  vet_id: string;
  name: string;
  description: string | null;
  price: number;
  duration_minutes: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface RevenueStats {
  total_revenue: number;
  monthly_revenue: number;
  avg_per_appointment: number;
  total_appointments_with_price: number;
}

export interface RevenueByServiceItem {
  service_name: string;
  total_revenue: number;
  appointment_count: number;
}

export interface RevenueByServiceResponse {
  items: RevenueByServiceItem[];
  total_revenue: number;
}

// Public vet types (no auth required)
export interface PublicVet {
  id: string;
  name: string;
  slug: string;
  specialty: string | null;
  city: string | null;
  address: string | null;
  phone: string | null;
  image_url: string | null;
  rating_average: number;
  reviews_count: number;
  is_on_call: boolean;
  working_hours: Record<string, DayHours> | null;
}

export interface PublicVetDetail extends PublicVet {
  license_number: string | null;
  coordinates_lat: number | null;
  coordinates_lng: number | null;
  description: string | null;
}

export interface PublicVetListResponse {
  items: PublicVet[];
  total: number;
  page: number;
  page_size: number;
}

export interface PublicReview {
  id: string;
  rating: number;
  comment: string | null;
  created_at: string;
  owner_name: string;
}

export interface PublicReviewListResponse {
  items: PublicReview[];
  total: number;
  page: number;
  page_size: number;
}

export interface Reminder {
  id: string;
  pet_id: string;
  pet_name: string | null;
  vet_id: string;
  vet_name: string | null;
  pet_owner_id: string;
  type: 'vaccination' | 'checkup' | 'medication' | 'custom';
  title: string;
  message: string | null;
  due_date: string;
  reminder_date: string;
  is_sent: boolean;
  is_dismissed: boolean;
  created_at: string;
}

File: vetly-back/app/schemas/vet.py
"""
Vet schemas for request/response validation
"""

from uuid import UUID
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, EmailStr, ConfigDict, Field


class Shift(BaseModel):
    """A single work shift (morning or afternoon)"""
    open: str = Field(..., pattern=r'^\d{2}:\d{2}$')
    close: str = Field(..., pattern=r'^\d{2}:\d{2}$')


class DayHours(BaseModel):
    """Working hours for a single day with two optional shifts"""
    closed: bool = False
    morning: Shift | None = None
    afternoon: Shift | None = None


class WorkingHours(BaseModel):
    """Weekly working hours"""
    monday: DayHours | None = None
    tuesday: DayHours | None = None
    wednesday: DayHours | None = None
    thursday: DayHours | None = None
    friday: DayHours | None = None
    saturday: DayHours | None = None
    sunday: DayHours | None = None


class VetBase(BaseModel):
    """Shared vet fields"""
    name: str
    email: EmailStr
    specialty: str
    phone: str
    address: str
    city: str
    description: str | None = None
    image_url: str | None = None


class VetResponse(VetBase):
    """Vet response returned from API"""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    license_number: str
    coordinates_lat: Decimal | None = None
    coordinates_lng: Decimal | None = None
    hours: dict | None = None
    is_on_call: bool
    is_verified: bool
    email_verified: bool
    rating_average: Decimal
    reviews_count: int
    created_at: datetime
    updated_at: datetime


class VetListResponse(BaseModel):
    """Paginated list of vets"""
    items: list[VetResponse]
    total: int
    page: int
    page_size: int


class VetUpdateRequest(BaseModel):
    """Vet profile update request"""
    name: str | None = Field(None, min_length=2, max_length=255)
    specialty: str | None = Field(None, min_length=2, max_length=255)
    phone: str | None = Field(None, min_length=5, max_length=50)
    address: str | None = Field(None, min_length=5, max_length=500)
    city: str | None = Field(None, min_length=2, max_length=100)
    description: str | None = Field(None, max_length=2000)
    image_url: str | None = Field(
        None,
        max_length=500,
        pattern=r"^/uploads/(vet|pet|pet_cover|owner)/[a-f0-9]{32}\.(jpg|png)$",
    )
    coordinates_lat: Decimal | None = Field(None, ge=-90, le=90)
    coordinates_lng: Decimal | None = Field(None, ge=-180, le=180)


class VetHoursUpdateRequest(BaseModel):
    """Vet working hours update request"""
    hours: WorkingHours


class OnCallToggleRequest(BaseModel):
    """On-call status toggle request"""
    is_on_call: bool


class OnCallVetsResponse(BaseModel):
    """List of on-call vets (unpaginated)"""
    items: list[VetResponse]
    count: int


class AvailableSlotsResponse(BaseModel):
    """Response for available time slots"""
    date: str
    vet_id: UUID
    slots: list[str]
